From fbb883e688c24973ecbdf799ef6c741d37ee1b86 Mon Sep 17 00:00:00 2001
From: jasjuang <jasjuang@gmail.com>
Date: Mon, 15 Oct 2018 09:46:19 -0700
Subject: [PATCH] add support for latest Turing gpu and cuda 10

---
 cmake/pcl_find_cuda.cmake | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/cmake/pcl_find_cuda.cmake b/cmake/pcl_find_cuda.cmake
index b6a0987..cfe51cf 100644
--- a/cmake/pcl_find_cuda.cmake
+++ b/cmake/pcl_find_cuda.cmake
@@ -42,8 +42,10 @@ if(CUDA_FOUND)
 
   # Find a complete list for CUDA compute capabilities at http://developer.nvidia.com/cuda-gpus
 
-  if(NOT ${CUDA_VERSION_STRING} VERSION_LESS "9.0")
-    set(__cuda_arch_bin "3.0 3.5 5.0 5.2 5.3 6.0 6.1 7.0")
+  if(NOT ${CUDA_VERSION_STRING} VERSION_LESS "10.0")
+    set(__cuda_arch_bin "3.0 3.5 5.0 5.2 5.3 6.0 6.1 7.0 7.2 7.5")
+  elseif(NOT ${CUDA_VERSION_STRING} VERSION_LESS "9.0")
+    set(__cuda_arch_bin "3.0 3.5 5.0 5.2 5.3 6.0 6.1 7.0 7.2")
   elseif(NOT ${CUDA_VERSION_STRING} VERSION_LESS "8.0")
     set(__cuda_arch_bin "2.0 2.1(2.0) 3.0 3.5 5.0 5.2 5.3 6.0 6.1")
   elseif(NOT ${CUDA_VERSION_STRING} VERSION_LESS "6.5")
-- 
2.7.4

